### ISSUE TITLE
bug: Wrong SELinux context causing crashes

### ISSUE THREAD

**Issue by Hakimi0804** [ReVanced/revanced-manager#510]
### Type

Error at runtime

### Bug description

Upon every reboot, the bind-mounted APK have a wrong SELinux context causing it to fail to start properly.

### Steps to reproduce

1. Patch the APK
2. Install via root method
3. Reboot
4. Launch the app [crashes here]

### Android version

Android 12 (S)

### Manager version

0.0.36

### Target package name

com.google.android.apps.youtube.music (YouTube Music)

### Target package version.

5.31.50

### Installation type

Root

### Patches selected.

Background Play 0.0.1
Exclusive Audio Playback 0.0.1
Hide Get Premium 0.0.1
Music Video Ads 0.0.1

### Device logs (exported using Manager settings).

```shell
manager log output: (Seems to contain nothing helpful)
revanced-manager_logcat_20221114115406313807.log
--------- beginning of main
11-14 11:53:24.730 16432 19332 D SurfaceComposerClient: VRR [FRTC] client handle [bufferId:18446744073709551615 framenumber:0] [ffffffff, ffffffff]
11-14 11:53:24.730 16432 19332 D SurfaceComposerClient: VRR [FRTC] client handle [bufferId:18446744073709551615 framenumber:0] [ffffffff, ffffffff]
11-14 11:53:25.715 16432 16432 V AutofillManager: requestHideFillUi(null): anchor = null
11-14 11:53:26.272 16432 16432 V AutofillManager: requestHideFillUi(null): anchor = null
11-14 11:53:27.252 16432 16432 V AutofillManager: requestHideFillUi(null): anchor = null


Relevant output from logcat:
11-14 07:43:00.855  2856  2922 E ApkChecksums: Signature verification error
11-14 07:43:00.855  2856  2922 E ApkChecksums: android.content.pm.PackageParser$PackageParserException: Failed to collect certificates from /data/app/~~YvoWicKIVzyPe1Cp4ChPvw==/com.google.android.youtube-ZTEIm4pta1A-LERUGrUXCA==/base.apk using APK Signature Scheme v3
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.util.apk.ApkSignatureVerifier.verifyV3Signature(ApkSignatureVerifier.java:329)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.util.apk.ApkSignatureVerifier.verifyV3AndBelowSignatures(ApkSignatureVerifier.java:162)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.util.apk.ApkSignatureVerifier.verifySignaturesInternal(ApkSignatureVerifier.java:154)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at com.android.server.pm.ApkChecksums.extractHashFromV2V3Signature(ApkChecksums.java:666)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at com.android.server.pm.ApkChecksums.getAvailableApkChecksums(ApkChecksums.java:404)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at com.android.server.pm.ApkChecksums.getChecksums(ApkChecksums.java:320)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at com.android.server.pm.PackageManagerService.lambda$requestChecksumsInternal$4$PackageManagerService(PackageManagerService.java:6723)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at com.android.server.pm.PackageManagerService$$ExternalSyntheticLambda42.run(Unknown Source:16)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.os.Handler.handleCallback(Handler.java:938)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.os.Handler.dispatchMessage(Handler.java:99)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.os.Looper.loopOnce(Looper.java:233)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.os.Looper.loop(Looper.java:344)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.os.HandlerThread.run(HandlerThread.java:67)
11-14 07:43:00.855  2856  2922 E ApkChecksums: Caused by: java.io.FileNotFoundException: /data/app/~~YvoWicKIVzyPe1Cp4ChPvw==/com.google.android.youtube-ZTEIm4pta1A-LERUGrUXCA==/base.apk: open failed: EACCES (Permission denied)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at libcore.io.IoBridge.open(IoBridge.java:574)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at java.io.RandomAccessFile.<init>(RandomAccessFile.java:289)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at java.io.RandomAccessFile.<init>(RandomAccessFile.java:152)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.util.apk.ApkSignatureSchemeV3Verifier.verify(ApkSignatureSchemeV3Verifier.java:119)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.util.apk.ApkSignatureSchemeV3Verifier.unsafeGetCertsWithoutVerification(ApkSignatureSchemeV3Verifier.java:114)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at android.util.apk.ApkSignatureVerifier.verifyV3Signature(ApkSignatureVerifier.java:309)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  ... 12 more
11-14 07:43:00.855  2856  2922 E ApkChecksums: Caused by: android.system.ErrnoException: open failed: EACCES (Permission denied)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at libcore.io.Linux.open(Native Method)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at libcore.io.BlockGuardOs.open(BlockGuardOs.java:274)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  at libcore.io.IoBridge.open(IoBridge.java:560)
11-14 07:43:00.855  2856  2922 E ApkChecksums:  ... 17 more
11-14 07:43:00.856  2856  2922 E ApkChecksums: Error reading /data/app/~~YvoWicKIVzyPe1Cp4ChPvw==/com.google.android.youtube-ZTEIm4pta1A-LERUGrUXCA==/base.apk to compute hash.
```

### Screenshots or videos

https://user-images.githubusercontent.com/67860246/201572240-cbd9e743-f95f-4650-a834-0fad07310d62.mp4

### Solution

1. From root shell, run: `restorecon -FRv /data/app`
2. Relaunch the app

### Additional context

Maybe the boot script installed by the manager could run a restorecon after the bind-mount?

### Acknowledgements

- [X] I have searched the existing issues and this is a new and no duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I filled out all of the requested information in this issue properly.
- [X] The issue is related solely to the ReVanced Manager

**Comment by ILoveOpenSourceApplications** [comment]
What's the state of this bug? Does it still exist? If not this can be closed

**Comment by Ten0** [comment]
I do have the same issue for any patched app installed via root. It may be tied to #519.

I also get the same error messages specifically when it doesn't work (the apps don't start anymore, which is quite a bother when the app is my application launcher):
```
03-31 03:14:19.120  1684  2089 I ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.HOME] flg=0x10200000 cmp=com.teslacoilsw.launcher/.NovaLauncher (has extras)} from uid 1000
03-31 03:14:19.122  1684  2089 W ziparchive: Unable to open '/data/app/com.teslacoilsw.launcher-tA9mTbXduHbjZwZ5uBuopQ==/base.apk': Permission denied
```

I can also confirm that `restorecon -FRv /data/app` gives back (only, so that really just fixes here and the rest is properly setup):
```
SELinux: Loaded file_contexts
SELinux:  Relabeling /data/app/com.teslacoilsw.launcher-tA9mTbXduHbjZwZ5uBuopQ==/base.apk from u:object_r:shell_data_file:s0 to u:object_r:apk_data_file:s0.
SELinux:  Relabeling /data/app/com.google.android.youtube-YGXQGvB_dzin8r1DZwdJOQ==/base.apk from u:object_r:shell_data_file:s0 to u:object_r:apk_data_file:s0.
```
and fixes a bunch of issues on both apps (nova not starting and losing its icon and app name, youtube starting very slow and losing its icon and app name...)

**Comment by Hakimi0804** [comment]
> What's the state of this bug? Does it still exist? If not this can be closed

Hi, sorry for the late reply. I can no longer reproduce this bug (as in the SELinux context is now correct and it's no longer crashing). I guess it can be closed(?)

@Ten0 Do you still have this issue?

**Comment by Hakimi0804** [comment]
Ok, it seems like [this commit](https://github.com/revanced/revanced-manager/commit/6ef716b0e4ef711a9b1ed04d822143c705d9ad0e) sets the correct context.

**Comment by Ten0** [comment]
> as in the SELinux context is now correct

My previous message was already with my phone on the latest release of RevancedManager (0.0.57, https://github.com/revanced/revanced-manager/releases), so nothing has changed since then for me, and I still resort to typing your command at every boot. (I also have another issue where Youtube stops working entirely and half of the system seems to believe that it's uninstalled until reboot, so the combination of both is really annoying)

**Comment by Ten0** [comment]
that commit was already included in a release prior to you opening the issue

**Comment by Hakimi0804** [comment]
> it seems that this commit was already included in a release prior to you opening the issue

Right, then I've no clue why could I no longer reproduce this. Anyways, we should leave this opened in the meantime

**Comment by Hakimi0804** [comment]
> IIUC the fix is just adding a line around here?
> 
> https://github.com/revanced/revanced-manager/blob/866a6e4a4480e837803263a539c9cf7792bdd943/lib/services/root_api.dart#L147

This would do it tbh

**Comment by mrksbk** [comment]
My [Bug Report](https://github.com/revanced/revanced-patches/issues/1955#event-9073550034) has been closed and linked here.

The problem that the patched YouTube **18.08.37** is no longer usable after a restart is (in my opinion) related to either the latest Patches **2.169.0** or the ReVanced Integration **0.103.0** and only affects the Root area. With MicroG, on the other hand, there are no problems with the versions mentioned above.

It also makes no difference which version of the Manager is used (**0.57 or 0.1.1**)

I also tested the Magisk Module with Youtube **18.08.37** or Revanced Extended with a higher version number of YouTube, which all lead to the same problem.

Currently I'm using YouTube ReVanced built with CLI with YouTube **18.05.40**, Patches **2.167.1** and Revanced Integration **0.102.0** with no problems.

**Comment by Hakimi0804** [comment]
@Ten0 So I found this line in /system/etc/init/hw/init.rc, which restores SELinux of /data recursively. I did notice that the SELinux bug stopped happening after a software (hence I couldn't reproduce it anymore), but I'm not sure if this line was added on that ota, or it's just a coincidence.

![Screenshot_2023-04-23-23-29-19-37_dac7cc7571c39b392df64923967cf7da](https://user-images.githubusercontent.com/67860246/233849426-eb2e322f-1bfa-462e-8bfb-4c59f26f4a64.jpg)

Can you check if you already have this line? (`su -c "grep restorecon /system/etc/init/hw/init.rc"`)

**Comment by Ten0** [comment]
I don't have it but I'm not super up to date on my Android version.
I've got a bunch of `.rc` files in `/system/etc/init`, none of them containing restorecon.
(except a wifi one that applies only to a wifi-related folder)

For now I've manually patched the generated files, but it still looks like we should do the fix described at https://github.com/revanced/revanced-manager/issues/510#issuecomment-1500341019 so that we don't rely on having by chance something else calling `restorecon` on `/data/app`.
